### ISSUE TITLE
checker-framework.groovy (differences): false positives when running checker-framework-nullness-optional-interning

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#12221]
When running `groovy ./.ci/checker-framework.groovy checker-framework-nullness-optional-interning` on a clean master, I get:
````
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  12:34 min
[INFO] Finished at: 2022-09-20T20:47:21-04:00
[INFO] ------------------------------------------------------------------------
New surviving error(s) found:

File Name: "checkstyle\src\main\java\com\puppycrawl\tools\checkstyle\Checker.java"
Specifier: "argument"
Message: "incompatible argument for parameter moduleId of Violation."
Line Contents: "new String[] {ioe.getMessage()}, null, getClass(), null));"
Line Number: 342
Details: found   : null (NullType)
....
````

I get 308 new errors and 643 unnecessary errors.

I a running on Windows.

**Comment by nrmancuso** [comment]
@rnveach please share your groovy version

**Comment by nrmancuso** [comment]
Similar issue here, on linux:
```
➜  checkstyle git:(master) java --version
openjdk 18.0.2 2022-07-19
OpenJDK Runtime Environment (build 18.0.2+0)
OpenJDK 64-Bit Server VM (build 18.0.2+0, mixed mode)
➜  checkstyle git:(master) groovy --version
Groovy Version: 2.5.18 JVM: 18.0.2 Vendor: Oracle Corporation OS: Linux

```

Seeing as how checker relies on javac, it is possible that compiler warnings/errors can be different with each java version/vendor.

**Comment by Vyom-Yadav** [comment]
I am on it and will share the reason by tomorrow.

**Comment by Vyom-Yadav** [comment]
New errors coming up could be due to version/vendor of javac, I sent a fix for proper equals and hashCode contract in #12223, please check after applying that patch.

**Comment by Vyom-Yadav** [comment]
> > I sent a fix for proper equals and hashCode contract in #12223, please check after applying that patch.
> 
> No real difference on my end.

Please try with the version and distribution used in CI, it looks to be version dependent.

**Comment by nrmancuso** [comment]
> > > I sent a fix for proper equals and hashCode contract in #12223, please check after applying that patch.
> > 
> > 
> > No real difference on my end.
> 
> Please try with the version and distribution used in CI, it looks to be version dependent.

Please provide details about version and distribution

**Comment by rnveach** [comment]
> Please try with the version and distribution used in CI

So have we confirmed this tool is directly tied to a specific compiler and if something changes in our CI or our local is different, we can expect differences in violations?

If we can narrow down a specific example, we should report this issue to the maintainers.

This doesn't seem like a very stable tool if we force users to only use a specific compiler. Even for pitest, I am able to to reproduce things on my local.

====

I compared 2 reported differences and here is what I am seeing:

````
New surviving error(s) found:

File Name: "checkstyle\src\main\java\com\puppycrawl\tools\checkstyle\Checker.java"
Specifier: "argument"
Message: "incompatible argument for parameter moduleId of Violation."
Line Contents: "new String[] {ioe.getMessage()}, null, getClass(), null));"
Line Number: 342
Details: found   : null (NullType)
````

````
File Name: "checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/Checker.java"
Specifier: "argument"
Message: "incompatible argument for parameter moduleId of Violation."
Line Contents: "new String[] {ioe.getMessage()}, null, getClass(), null));"
Details: found   : null (NullType)
         required: @Initialized @NonNull String
````

File name is different by slashes, `\` versus `/`.
One has line number and not the other.
One says `null (NullType)` while the other says `null (NullType) required: @Initialized @NonNull String`.

**Comment by rnveach** [comment]
> File name is different by slashes, \ versus /.

Numbers changed when fixing this.

Before was `308 new errors and 643 unnecessary errors` but is now `114 new errors and 449 unnecessary errors`.

**Comment by rnveach** [comment]
> One says null (NullType) while the other says null (NullType) required: @Initialized @NonNull String.
> I did a search and confirmed nothing in the new is showing required: .

Comparing CI:
````
2022-09-22T15:37:00.0077585Z [WARNING] /home/runner/work/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java:[290,74] [argument] incompatible argument for parameter customMessage of Violation.
2022-09-22T15:37:00.0078097Z   found   : null (NullType)
2022-09-22T15:37:00.0078376Z   required: @Initialized @NonNull String
````
To Local:
````
M:\checkstyleWorkspaceEclipse\checkstyle\src\main\java\com\puppycrawl\tools\checkstyle\PackageObjectFactory.java:[290,56] [argument] incompatible argument for parameter moduleId of Violation.
  found   : null (NullType)
[WARNING] 
  required: @Initialized @NonNull String
````

It could be a parsing error on our end as there is extra `[WARNING]` text, I don't know this program well.

**Comment by Vyom-Yadav** [comment]
I will add support for windows paths too in the suppression model, and won't modify it like other suppression files where we have:
```
<fileName>checkstyle[\\/]src[\\/]main[\\/]java[\\/]com[\\/]puppycrawl[\\/]tools[\\/]checkstyle[\\/]Checker.java</fileName>
```
Instead, I would internally replace windows path separator with Linux ones while creating mutations.

**Comment by Vyom-Yadav** [comment]
> Please provide details about version and distribution

From CI

```
$ java --version

openjdk 11.0.16 2022-07-19 LTS
OpenJDK Runtime Environment Zulu11.58+15-CA (build 11.0.16+8-LTS)
OpenJDK 64-Bit Server VM Zulu11.58+15-CA (build 11.0.16+8-LTS, mixed mode)
```

```
$ groovy --version

Groovy Version: 2.4.17 JVM: 11.0.16 Vendor: Azul Systems, Inc. OS: Linux
```

**Comment by romani** [comment]
@Vyom-Yadav , please help to resolve this issue.

We need to change parser to parser output more reliably to skip lines with "WARNING"

**Comment by nrmancuso** [comment]
I am looking into this

**Comment by rnveach** [comment]
This issue is approved, but will no longer be used as a base for PRs. New issues are needed for each item to work on.

**Comment by nrmancuso** [comment]
Experience so far with Checker shows that it is very fragile due to reliance on compiler version and OS. Now that we have clearly identified some problems, I suggest we:

1. Create official "parent" issue for Windows support
2. Create clear task list in parent issue that will result in level of Windows support we want to achieve
3. Close this issue

As part of tasks in parent issue, we should complete work at https://github.com/checkstyle/checkstyle/pull/12368 to keep at least one checker execution in CI to prevent regression in windows support.

As it stand now, we have details and context scattered across issues and PRs, let's consolidate all knowledge into parent issue.

**Comment by Vyom-Yadav** [comment]
According to current knowledge, supporting windows OS is not possible because the content of logs is different for windows and Linux. See https://github.com/checkstyle/checkstyle/pull/12366#issuecomment-1304441297

I created #12695 with more details, can we close this?

**Comment by rnveach** [comment]
@Vyom-Yadav https://github.com/checkstyle/checkstyle/pull/12366#issuecomment-1304585080
> please report this issue to Checker as it is their problem
> please report a separate issue to Checker that we would like a way for them to create a report file so we can completely drop maven output parsing

Have these 2 issues been raised to Checker? If so, please provide their links in their respective Checkstyle issues and let me know when that is done.

**Comment by romani** [comment]
@Vyom-Yadav , ping

**Comment by Vyom-Yadav** [comment]
I will complete pending items on weekend :)

**Comment by romani** [comment]
@Vyom-Yadav, please help us create issues on Checker and we can close all issues in this summer project, and close summer project at all.

**Comment by Vyom-Yadav** [comment]
Reported to Checker Framework maintainers at https://github.com/typetools/checker-framework/issues/5665

**Comment by rnveach** [comment]
https://github.com/checkstyle/checkstyle/issues/12221#issuecomment-1411094125
> please report a separate issue to Checker that we would like a way for them to create a report file so we can completely drop maven output parsing

Has this one been raised too?

**Comment by Vyom-Yadav** [comment]
> [#12221 (comment)](https://github.com/checkstyle/checkstyle/issues/12221#issuecomment-1411094125)
> 
> > please report a separate issue to Checker that we would like a way for them to create a report file so we can completely drop maven output parsing
> 
> Has this one been raised too?

Reported at https://github.com/typetools/checker-framework/issues/5666

**Comment by rnveach** [comment]
Thank you. Please make a similar issue for it in our repo to connect the 2 and then I think we can close this issue.

**Comment by romani** [comment]
moved to https://github.com/checkstyle/checkstyle/issues/12754
